File: src/api/resource-assignments.ts
import { RequestFN } from "~/core/request-function";
import { decodeAssignment } from "~/decoders/assignment";
import { TabLocation, type Assignment, type Resource, type SessionHandle } from "~/models";

/**
 * Retrieve assignments from a resource.
 */
export const resourceAssignments = async (session: SessionHandle, resource: Resource): Promise<Array<Assignment>> => {
  // prevent doing an unnecessary request.
  if (!resource.haveAssignment) return [];

  const request = new RequestFN(session, "donneesContenusCDT", {
    _Signature_: { onglet: TabLocation.Resources },

    donnees: {
      pourTAF: true,
      cahierDeTextes: { N: resource.id }
    }
  });

  const response = await request.send();

  return response.data.donnees.ListeCahierDeTextes.V[0].ListeTravailAFaire.V
    .map((assignment: any) => decodeAssignment(assignment, session));
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/resource-assignments.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeAssignment
import ink.literate.pawnote.models.Assignment
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

/**
 * Retrieve assignments from a resource.
 */
suspend fun resourceAssignments (session: SessionHandle, resourceID: String): List<Assignment> {
    val request = RequestFN(session.information, "donneesContenusCDT", Json.encodeToString(
        buildJsonObject {
            putJsonObject("_Signature_") {
                put("onglet", TabLocation.Resources.code)
            }

            putJsonObject("donnees") {
                put("pourTAF", true)
                putJsonObject("cahierDeTextes") {
                    put("N", resourceID)
                }
            }
        }
    ))

    val response = request.send()

    return Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject["ListeCahierDeTextes"]!!.jsonObject["V"]!!
        .jsonArray[0].jsonObject["ListeTravailAFaire"]!!.jsonObject["V"]!!
        .jsonArray.map { decodeAssignment(it.jsonObject, session.information) }
}